module/zfs/vdev_raidz.c

```diff
@@ -135,6 +141,206 @@
+ *  |    13|    14|    15|    16|         |    16|    17|    18|p   19|    20|
+ *  +------+------+------+------+         +------+------+------+------+------+
+ *  |      |      |      |      |         |      |      |      |      |      |
+ *  |  F   |  F   |  G   |  G   |         |  G   |  G   |  H   |  H   |  H   |
+ *  |    17|    18|    19|    20|         |    21|    22|    23|    24|    25|
+ *  +------+------+------+------+         +------+------+------+------+------+
+ *  |      |      |      |      |         |      |      |      |      |      |
+ *  |  G   |  G   |  H   |  H   |         |  H   |  I   |  I   |  J   |  J   |
+ *  |    21|    22|    23|    24|         |    26|    27|    28|    29|    30|
+ *  +------+------+------+------+         +------+------+------+------+------+
+ *  |      |      |      |      |         |      |      |      |      |      |
+ *  |  H   |  H   |  I   |  I   |         |  J   |  J   |      |      |  K   |
+ *  |    25|    26|    27|    28|         |    31|    32|    33|    34|    35|
+ *  +------+------+------+------+         +------+------+------+------+------+
+ *
+ * This reflow approach has several advantages. There is no need to read or
+ * modify the block pointers or recompute any block checksums.  The reflow
+ * doesn’t need to know where the parity sectors reside. We can read and write
+ * data sequentially and the copy can occur in a background thread in open
+ * context. The design also allows for fast discovery of what data to copy.
+ *
+ * The VDEV metaslabs are processed, one at a time, to copy the block data to
+ * have it flow across all the disks. The metaslab is disabled for allocations
+ * during the copy. As an optimization, we only copy the allocated data which
+ * can be determined by looking at the metaslab range tree. During the copy we
+ * must maintain the redundancy guarantees of the RAIDZ VDEV (e.g. parity count
+ * of disks can fail). This means we cannot overwrite data during the reflow
+ * that would be needed if a disk is lost.
+ *
+ * After the reflow completes, all newly-written blocks will have the new
+ * layout, i.e., they will have the parity to data ratio implied by the new
+ * number of disks in the RAIDZ group.
+ *
+ * Even though the reflow copies all the allocated space, it only rearranges
+ * the existing data + parity. This has a few implications about blocks that
+ * were written before the reflow completes:
+ *
+ *  - Old blocks will still use the same amount of space (i.e., they will have
+ *    the parity to data ratio implied by the old number of disks in the RAIDZ
+ *    group).
+ *  - Reading old blocks will be slightly slower than before the reflow, for
+ *    two reasons. First, we will have to read from all disks in the RAIDZ
+ *    VDEV, rather than being able to skip the children that contain only
+ *    parity of this block (because the parity and data of a single block are
```

Comment:
Has there already been some form of test case created to quantify this impact in any way by chance? I've looked and been unable to find one up to now.

I'm thinking about many of 'majority read IO' scenarios, where much of the data may be considered somewhat stagnant but is read often by users/applications (file hosting, etc). Datasets like these typically grow quite slowly, possibly making the use of the expansion feature (once available) more likely, and having at least some kind of understanding of the level of performance impact could help with their decision making process.

In the event a test case doesn't exist, I'm trying to brainstorm how one might go about creating one that would even cover a significant portion of scenarios one might find themselves in... For example, I'd imagine this would play more in to expanding a VDEV from 5 drives 6 (all other things being equal) than going from 14 -> 15 (as I understand it at least?), as parity is a smaller percentage of the total raw VDEV capacity. The extent to which though, I think having at least some base understanding of what can be expected would be useful...

If this convo shouldn't continue here, please just let me know and I'll move it to discussions - it seemed at least related enough to send a first pass note anyway, but apologies if misdirected!

Reply:
[Slide 27](https://docs.google.com/presentation/d/1FeQgEwChrtNQBHfWSNsPK3Y53O5BnPh3Cz5nRa5GAQY/edit#slide=id.gdc016e85d2_0_2) of the presentation linked above has some performance results from a few years back; it shouldn't have changed much since then.  This was trying to examine the worst case, which is indeed a small number of drives (3->4 in my test), with 512-byte sector size.  Sequential read performance decreased by about 10%, from 2600 to 2300 MB/s.

It should be possible to add test cases for this to the performance section of the ZFS Test Suite.  However, I'm not sure if that is being run as part of any public test automation.

If you're able to do some performance testin in your lab, we'd be happy to hear the results!

Reply:
The current tests are functional and stress.  As Matt mentioned, there is a performance section in the test suite that could be leveraged (`tests/zfs-tests/tests/perf`) but that is not something we plan to do for this PR.